rfcs/0157-more-granular-control.md

```diff
@@ -0,0 +1,241 @@
+nix-store --import < file
+```
+
+The purpose of remote builder shouldn't be exactly to serve the built file immediately, maybe to store it in some static-storage, and shutdown, the cache serving should be made accessible using stateless http server, which can happen if we can serve a folder through http.
+
+Articles confuse the binary cache with closure copying, such as [this one](https://nix-tutorial.gitlabpages.inria.fr/nix-tutorial/deploying.html), one of top results in Google with "setting up a binary cache nix" query.
+
+Hence there shouldn't be need of third party package in order to serve something, unnecessary complexity & abstractions should be reduced to fit a simple custom http file server to act as binary cache.
+
+```bash
+# Something like:
+nix store setup > /nix/store/nix-cache-info
+
+python3 -m http.server -d /nix/store
+```
+
+An alternative way could be use of a flake to build isolated directory with the filtered store path references only, or even an option to have upstream filtered paths (that are not in https://cache.nixos.org). Something like:
+
+```nix
+{
+  # ...
+  outputs = args: {
+    packages.x86_64-linux.my-binary-cache = some-binary-cache-setup-function {
+      packages = [ custom-package-1 custom-package-2 ];
+      # ...
+    };
+  };
+}
+
+# nix build '.#my-binary-cache'
+# python3 -m http.server -d result
+```
+
+Again, an indirect result of flake enforcement.
+
+### (optional) stdenv improvements
+
+Build requirement of C/C++ should be opt-in, it isn't generally true that everything requires C/C++ compiler and header/ldflag setup.
+
+```nix
+stdenv.mkDerivation {
+  includeCoreutils = true;
+  includeC = true;
+  # ...
+}
+```
+
+And most of the time usage of the `builtins.derivation` is far from practical, examples shown in Prior Arts section below.
+
+
+# Drawbacks
+[drawbacks]: #drawbacks
+
+If everything is made opt-in, it may result in a slightly longer initial configuration file, although way more explicit.
+
+There are no perceived drawbacks in flake enforcement / nix-cli improvements.
+
+
+# Alternatives
+[alternatives]: #alternatives
+
+### What other designs have been considered? What is the impact of not doing this?
+
+Not too sure about alternatives, let's discuss potential alternatives (if there are).
+
+If left unimplemented however, abstraction will keep growing and complexity will continue to add up.
+
+
+# Prior art
+[prior-art]: #prior-art
+
+### Remote Builds
+
+* [reference #1 (reddit)](https://www.reddit.com/r/NixOS/comments/l5221x/comment/gm9xre8).
+* [reference #2 (gist)](https://gist.github.com/danbst/09c3f6cd235ae11ccd03215d4542f7e7).
+* [reference #3 (LnL7/nix-docker) - Setting up a docker container as a remote builder](https://github.com/LnL7/nix-docker/tree/master#running-as-a-remote-builder).
+* [reference #4 (DeterminateSystems/magic-nix-cache#24)](https://github.com/DeterminateSystems/magic-nix-cache/issues/24).
+
+### Directly using builtins.derivation
```

Comment:
IMnsHO, `builtins.derivation` is similar to `ninja.build`: too low-level to be used directly by human beings.